src/raft/raft.go

```diff
@@ -41,25 +46,319 @@ type ApplyMsg struct {
 }
 
+func (rf *Raft) isLeader() (isLeader bool) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	isLeader = rf.role == 2
+	return
+}
+
+func (rf *Raft) isCandidate() (isCandidate bool) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	isCandidate = rf.role == 1
+	return
+}
+
+func (rf *Raft) isFollower() (isFollower bool) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	isFollower = rf.role == 0
+	return
+}
+
+func (rf *Raft) makeLeader() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.role = 2
+	return
+}
+
+func (rf *Raft) makeCandidate() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.role = 1
+	return
+}
+
+func (rf *Raft) makeFollower() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.role = 0
+	return
+}
+
+func (rf *Raft) incTerm() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm++
+	return
+}
+
+func (rf *Raft) decTerm() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm--
+	return
+}
+
+func (rf *Raft) setTerm(term int) {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.currentTerm = term
+	return
+}
+
+func (rf *Raft) getMajority() int {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	return rf.majority
+
+}
+
+func (rf *Raft) setLastContact() {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	rf.lastContact = time.Now()
+}
+
+func (rf *Raft) LastContact() time.Time {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	contact := rf.lastContact
+	return contact
+}
+
+func (rf *Raft) getID() (ID int) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	ID = rf.me
+	return
+}
+
+func (rf *Raft) getTerm() (term int) {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	term = rf.currentTerm
+	return
+}
+
 // return currentTerm and whether this server
 // believes it is the leader.
 func (rf *Raft) GetState() (int, bool) {
-
-	var term int
-	var isleader bool
+	term := rf.getTerm()
+	isLeader := rf.isLeader()
 	// Your code here.
-	return term, isleader
+	return term, isLeader
+}
+
+// This is our main loop. It checks the state of the peer and drops down into the applicable loop
+func (rf *Raft) run() {
+	for {
+		switch {
+		case rf.killed:
+			DPrintf("Shutdown  🔫 ")
+			return
+		case rf.isFollower():
+			rf.runFollower()
+		case rf.isCandidate():
+			DPrintf("✋ Candidate state hit on %v", rf.me)
+			rf.runCandidate()
+		case rf.isLeader():
+			DPrintf("💪 leader state hit on %v", rf.me)
+			rf.runLeader()
+		}
+	}
+}
+
+func (rf *Raft) runFollower() {
+	heartbeatTimeout := randomTimeout(rf.heartbeatTimeout)
+
+	for {
+		DPrintf("👍🏻Follower state it on %v", rf.me)
+		select {
+		case <-rf.killCh:
+			close(rf.killCh)
+			DPrintf("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx------------------  Killed ")
+			return
+		case <-heartbeatTimeout:
+			heartbeatTimeout = randomTimeout(rf.heartbeatTimeout)
+			DPrintf("⏲ follower %v election timed out", rf.me)
+			lastContact := rf.LastContact()
+			if time.Now().Sub(lastContact) < rf.heartbeatTimeout {
+				continue
+			}
+			rf.makeCandidate()
+			return
+		case <-rf.heartBeatChan:
+			DPrintf("💖 follower %v received heartbeat", rf.me)
+			continue
+		}
+	}
+}
```

Comment:
Followers don't do all that much other than wait for heartbeat messags or become candidates if they have to wait too long.